### PR TITLE
separate axis internal API object and move responsibilities.

### PR DESCRIPTION
**Pull request by succhiello**
c3.js has global monkey patch to Chart and ChartInternal, and it seems making it difficult to improve modularity.

From now, I'll separate all modules from Chart and ChartInternal and add dependency management(probably use webpack).

This PR and following ones are very large architecture chagnes, so if it's unacceptable please feel free to decline it :smirk:. We will continue on our forked repository.

Changes has only syntactic changes, no logical change.

### CODE REVIEW

**Comment by aendrew** [comment]
Somebody get this man a medal. 

Seriously, @succhiello — while I haven't had time to look at it in too much detail, this is exactly what the C3 community is talking about right now. We've been directing modularity work around issue #351, and the discussion's come up again in #882. 

Furthermore, the axes have been causing me tonnes of grief lately — see #759, #895, #914, times/axisJS#42. I really need to have a right-aligned Y axis for times/axisJS, and tried to accomplish it in #899. I welcome anything that makes my next attempt to do so any less frustrating! :+1: 

It'll need a bit of review and @masayuki0812 has the final call, but just wanted to say it's so good to see that people are starting to address this aspect of C3.

**Comment by masayuki0812** [comment]
@succhiello Thank you so much for this PR. I agree with your concept and I'd like to merge basically, but can I request some modifications? I'll comment in the code, so I'd be happy if you check them.

File: c3.js [hunk 1]
```diff
@@ -154,6 +173,8 @@
         var $$ = this, d3 = $$.d3, config = $$.config;
         var defs, main, binding = true;
 
+        $$._axis = new _Axis($$);
```
Comment (masayuki0812) [inline c3.js]: Can you remove `_`? `$$.axis` seems better from the uniformity point of view (other internal variables do not have prefix `_`).

File: src/core.js [hunk 2]
```diff
@@ -149,6 +168,8 @@ c3_chart_internal_fn.initWithData = function (data) {
     var $$ = this, d3 = $$.d3, config = $$.config;
     var defs, main, binding = true;
 
+    $$._axis = new _Axis($$);
```
Comment (masayuki0812) [inline src/core.js]: Can you remove _? $$.axis seems better from the uniformity point of view (other internal variables do not have prefix _).

File: src/axis.js [hunk 3]
```diff
@@ -1,5 +1,12 @@
-c3_chart_internal_fn.initAxis = function () {
-    var $$ = this, config = $$.config, main = $$.main;
+function _Axis(owner) {
+    API.call(this, owner);
+}
+
+inherit(API, _Axis);
+
+_Axis.prototype.init = function init() {
```
Comment (masayuki0812) [inline src/axis.js]: Can you remove `_`? Just `Axis` seems better from the uniformity point of view (e.g. Chart, ChartInternal and API).

File: src/axis.js [hunk 4]
```diff
@@ -1,5 +1,12 @@
-c3_chart_internal_fn.initAxis = function () {
-    var $$ = this, config = $$.config, main = $$.main;
+function _Axis(owner) {
```
Comment (masayuki0812) [inline src/axis.js]: Can you remove `_`? Just `Axis` seems better from the uniformity point of view (e.g. Chart, ChartInternal and API).

File: src/axis.js [hunk 5]
```diff
@@ -399,7 +271,171 @@ c3_chart_internal_fn.redrawAxis = function (transitions, isHidden) {
...
     // rotate tick text if needed
     if (!config.axis_rotated && config.axis_x_tick_rotate) {
-        $$.rotateTickText($$.axes.x, transitions.axisX, config.axis_x_tick_rotate);
-        $$.rotateTickText($$.axes.subx, transitions.axisSubX, config.axis_x_tick_rotate);
+        _rotateTickText($$.axes.x, transitions.axisX, config.axis_x_tick_rotate);
+        _rotateTickText($$.axes.subx, transitions.axisSubX, config.axis_x_tick_rotate);
     }
 };
+
+function _getTickValues(tickValues, axis) {
```
Comment (masayuki0812) [inline src/axis.js]: Can you export these functions? As written in https://github.com/masayuki0812/c3/pull/929/files#r23514571, every function should be exported at this moment.

File: .jshintrc [hunk 6]
```diff
@@ -15,6 +15,7 @@
     "newcap": false,
     "loopfunc": true,
     "bitwise": false,
+    "validthis": true,
```
Comment (masayuki0812) [inline .jshintrc]: I think we can remove this if we export every internal axis function.

File: src/core.js (hunk 2 above)
Comment (succhiello) [inline src/core.js]: I think it's better code manner to add some prefix like underscore for emphasizing a intention  "this is internal and don't use from outside of the module!" and '$$' is ambiguous for that purpose...but I'll remove them anyway.

File: src/axis.js (hunk 3 above)
Comment (succhiello) [inline src/axis.js]: will do, but is there anything special reason why we use something like 'fn', instead of using ***.prototype directly?

**Comment by succhiello** [comment]
everything done!

**Comment by aendrew** [comment]
@succhiello You need to resolve the merge conflict. My guess is because the generated c3.js conflicts with a more recently generated one. We usually only build `c3.js` during releases; if you remove `c3.js` from the commit it should merge easier. :)

**Comment by succhiello** [comment]
@aendrew git reset --soft and unstage dist files.I think...dist files should be git-ignored completely and be added into the 'files' entry in package.json/bower.json.

**Comment by aendrew** [comment]
@masayuki0812 Any way I can help with verifying this PR so we can get it merged? It'll merge successfully at the moment, and any additional PRs to the axes will necessarily need to be merged after this (otherwise they'll conflict due to the scope of this one; already this will probably cause #904 to conflict.).

If I get a moment today I'll look through the updated commit and test it locally.

**Comment by masayuki0812** [comment]
I merged manually and it seems working well. Thanks! @succhiello 
@aendrew I'll look into another PRs later as well.

**Comment by aendrew** [comment]
Yay! Awesome. :+1:

**Pull request merged**